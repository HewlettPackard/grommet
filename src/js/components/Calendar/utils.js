
// Utility functions for the Calendar.
// Just what's needed to avoid having to include a dependency like momentjs.

const DAY_MILLISECONDS = 24 * 60 * 60 * 1000;

export const addDays = (date, days) =>
  (new Date(date.getTime() + (DAY_MILLISECONDS * days)));

export const subtractDays = (date, days) => addDays(date, -days);

export const addMonths = (date, months) => {
  const result = new Date(date);
  const years = Math.floor((date.getMonth() + months) / 12);
  result.setFullYear(date.getFullYear() + years);
  result.setMonth((date.getMonth() + months) % 12);
  return result;
};

export const subtractMonths = (date, months) => addMonths(date, -months);

export const sameDay = (date1, date2) => (
  date1.getFullYear() === date2.getFullYear() &&
  date1.getMonth() === date2.getMonth() &&
  date1.getDate() === date2.getDate()
);

export const sameDayOrAfter = (date1, date2) =>
  (date1.getFullYear() > date2.getFullYear() ||
    (date1.getFullYear() === date2.getFullYear() && (
      date1.getMonth() > date2.getMonth() || (
        date1.getMonth() === date2.getMonth() && (date1.getDate() >= date2.getDate())
      )
    ))
  );

export const sameDayOrBefore = (date1, date2) =>
  (date1.getFullYear() < date2.getFullYear() ||
    (date1.getFullYear() === date2.getFullYear() && (
      date1.getMonth() < date2.getMonth() || (
        date1.getMonth() === date2.getMonth() && (date1.getDate() <= date2.getDate())
      )
    ))
  );

export const daysApart = (date1, date2) =>
  Math.floor((date1.getTime() - date2.getTime()) / DAY_MILLISECONDS);

// betweenDates takes and array of two elements and checks if the
// supplied date lies between them, inclusive.
// returns 2 if exact match to one end, 1 if between, undefined otherwise
export const betweenDates = (date, dates) => {
  let result;
  if (dates) {
    const [from, to] = dates.map(d => new Date(d));
    if (sameDay(date, from) || sameDay(date, to)) {
      result = 2;
    } else if (sameDayOrAfter(date, from) && sameDayOrBefore(date, to)) {
      result = 1;
    }
  } else {
    result = 1;
  }
  return result;
};

// withinDates takes and array of string dates or 2 element arrays and
// checks whether the supplied date matches any string or is between
// any dates in arrays.
// returns 2 if exact match, 1 if between, undefined otherwise
export const withinDates = (date, dates) => {
  let result;
  if (dates) {
    if (Array.isArray(dates)) {
      dates.some((d) => {
        if (typeof d === 'string') {
          if (sameDay(date, new Date(d))) {
            result = 2;
          }
        } else {
          result = betweenDates(date, d);
        }
        return result;
      });
    } else if (sameDay(date, new Date(dates))) {
      result = 2;
    }
  }
  return result;
};
