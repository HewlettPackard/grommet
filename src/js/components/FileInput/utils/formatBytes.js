const SI_CONVERSION_FACTOR = 1000;
const IEC_CONVERSION_FACTOR = 1024;

const getBrowser = () => {
  const currentOS = ['Win', 'Linux', 'Mac'].find(
    (v) => window.navigator.userAgent.indexOf(v) >= 0,
  );

  return currentOS;
};

const defaultFormat = (size) => {
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  const factor = SI_CONVERSION_FACTOR;
  let index = 0;
  let num = size;
  while (num >= factor && index < units.length - 1) {
    num /= factor;
    index += 1;
  }
  return `${num.toFixed(1)} ${units[index]}`;
};

const windowsFormat = (size) => {
  const num = Math.ceil(size / IEC_CONVERSION_FACTOR);

  function formatNumber(number) {
    return number.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1.');
  }

  return `${formatNumber(num)} ${'KB'}`;
};

const makeFormatBytes = (browser) => (size) => {
  switch (browser) {
    case 'Win':
      return windowsFormat(size);
    default:
      return defaultFormat(size);
  }
};

const formatBytes = makeFormatBytes(getBrowser());

export { formatBytes };
