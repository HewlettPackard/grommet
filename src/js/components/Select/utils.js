import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { Box } from '../Box';
import { Button } from '../Button';
import { DropButton } from '../DropButton';
import { Text } from '../Text';
import { TextInput } from '../TextInput';
import {
  getHoverIndicatorStyle,
  selectedStyle,
  controlBorderStyle,
  normalizeColor,
} from '../../utils';

export const applyKey = (option, key) => {
  if (option === undefined) return undefined;
  if (typeof key === 'object') return applyKey(option, key.key);
  if (typeof key === 'function') return key(option);
  if (key !== undefined) return option[key];
  return option;
};

// position relative is so scroll can be managed correctly
export const OptionsBox = styled.div`
  position: relative;
  scroll-behavior: smooth;
  overflow: auto;
  outline: none;
`;

export const HiddenInput = styled.input`
  display: none;
`;

export const SelectOption = styled(Button)`
  ${(props) => props.selected && props.textComponent && selectedStyle}
  // applies theme.global.hover.background to the active
  // option for mouse and keyboard interactions
  ${(props) =>
    props.active &&
    getHoverIndicatorStyle(
      !props.children && !props.theme.select.options ? undefined : 'background',
      props.theme,
    )}
  display: block;
  width: 100%;
`;

export const SelectTextInput = styled(TextInput)`
  cursor: ${(props) => (props.defaultCursor ? 'default' : 'pointer')};
`;

export const StyledSelectDropButton = styled(DropButton)`
  ${(props) => !props.callerPlain && controlBorderStyle};
  ${(props) =>
    props.theme.select &&
    props.theme.select?.control &&
    props.theme.select?.control?.extend};
  ${(props) => props.open && props.theme.select?.control?.open};
`;

export const DefaultSelectTextInput = forwardRef(
  (
    { a11yTitle, disabled, id, name, placeholder, value, size, theme, ...rest },
    ref,
  ) => (
    <SelectTextInput
      a11yTitle={a11yTitle}
      // When Select is disabled, we want to show a default cursor
      // but not have disabled styling come from TextInput
      // Disabled can be a bool or an array of options to disable.
      // We only want to disable the TextInput if the control
      // button should be disabled which occurs when disabled
      // equals true.
      defaultCursor={disabled === true || undefined}
      focusIndicator={false}
      id={id ? `${id}__input` : undefined}
      name={name}
      ref={ref}
      {...rest}
      tabIndex="-1"
      type="text"
      placeholder={placeholder}
      plain
      readOnly
      value={value}
      size={size}
      theme={theme}
    />
  ),
);

export const EmptySearchOption = ({
  emptySearchMessage,
  selectOptionsStyle,
  theme,
}) => (
  <SelectOption
    key="search_empty"
    tabIndex="0"
    role="menuitem"
    hoverIndicator="background"
    disabled
    aria-live="polite"
  >
    <Box {...selectOptionsStyle}>
      <Text {...theme.select.container.text}>{emptySearchMessage}</Text>
    </Box>
  </SelectOption>
);

export const getOptionLabel = (index, options, labelKey) =>
  applyKey(options[index], labelKey);

export const getOptionValue = (index, options, valueKey) =>
  applyKey(options[index], valueKey);

export const checkDisabled = (
  index,
  disabled,
  disabledKey,
  options,
  valueKey,
) => {
  const option = options[index];
  let result;
  if (disabledKey) {
    result = applyKey(option, disabledKey);
  } else if (Array.isArray(disabled)) {
    if (typeof disabled[0] === 'number') {
      result = disabled.indexOf(index) !== -1;
    } else {
      const optionVal = getOptionValue(index, options, valueKey);
      result = disabled.indexOf(optionVal) !== -1;
    }
  }
  return result;
};

export const calcValuedValue = (value, valueKey) => {
  if (Array.isArray(value))
    return value.map((v) =>
      valueKey && valueKey.reduce ? v : applyKey(v, valueKey),
    );
  return valueKey && valueKey.reduce ? value : applyKey(value, valueKey);
};

export const changeEvent = (inputRef, nextValue) => {
  // Calling set value function directly on input because React library
  // overrides setter `event.target.value =` and loses original event
  // target fidelity.
  // https://stackoverflow.com/a/46012210
  const nativeInputValueSetter = Object.getOwnPropertyDescriptor(
    window.HTMLInputElement.prototype,
    'value',
  ).set;
  nativeInputValueSetter.call(inputRef.current, nextValue);
  const event = new Event('input', { bubbles: true });
  inputRef.current.dispatchEvent(event);
};

export const getSelectIcon = (icon, theme, open) => {
  let SelectIcon;
  switch (icon) {
    case false:
      break;
    case true:
    case undefined:
      SelectIcon =
        open && theme.select.icons.up
          ? theme.select.icons.up
          : theme.select.icons.down;
      break;
    default:
      SelectIcon = icon;
  }
  return SelectIcon;
};

// if labelKey is a function and valueLabel is not defined
// we should use the labelKey function to display the
// selected value
export const getDisplayLabelKey = (
  labelKey,
  allOptions,
  optionIndexesInValue,
  selectValue,
) => {
  const optionLabelKey = applyKey(
    allOptions[optionIndexesInValue[0]],
    labelKey,
  );
  if (
    !selectValue &&
    optionIndexesInValue.length === 1 &&
    typeof optionLabelKey === 'object'
  )
    return optionLabelKey;
  return undefined;
};

export const getIconColor = (theme) =>
  normalizeColor(theme.select.icons.color || 'control', theme);
