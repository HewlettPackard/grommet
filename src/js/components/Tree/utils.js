export const setupNodes = nodes =>
  nodes.map(node => ({
    name: node.name,
    childs: setupNodes(node.childs || []),
    opened: false,
  }));

export const isSameNode = (node1, node2) => node1 === node2;

export const changeNode = (positions, nodes) => {
  const [currentPosition, ...otherPositions] = positions;
  const currentNode = nodes[currentPosition];

  if (otherPositions.length) {
    return nodes.map(node =>
      isSameNode(node, currentNode)
        ? {
            ...node,
            childs: changeNode(otherPositions, node.childs),
          }
        : { ...node },
    );
  }

  return nodes.map(node =>
    isSameNode(node, currentNode)
      ? {
          ...node,
          opened: !node.opened,
        }
      : { ...node },
  );
};
