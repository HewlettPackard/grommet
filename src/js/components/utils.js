import { css } from 'styled-components';

export const focusStyle = css`
  border-color: ${
    props => (
      props.theme.global.focus.border.color ||
      props.theme.global.colors.accent[0]
    )
  };
  box-shadow: 0 0 1px 1px ${
    props => (
      props.theme.global.focus.border.color ||
      props.theme.global.colors.accent[0]
    )
  };
`;

export function parseMetricToInt(fontAsString) {
  return parseInt(fontAsString.replace(/[^0-9]/g, ''), 10);
}

export function findAllByType(component, type) {
  let matches = [];

  if (component.type === type) {
    matches.push(component);
  }

  if (component.children) {
    component.children.forEach((child) => {
      matches = matches.concat(findAllByType(child, type));
    });
  }

  return matches;
}

export default { findAllByType, focusStyle, parseMetricToInt };
