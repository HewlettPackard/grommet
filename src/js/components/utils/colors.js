
export const colorForName = (name, theme) => {
  const [kind, index] = name.split('-');
  const colorSet = theme.global.colors[kind];
  let color;
  if (Array.isArray(colorSet)) {
    color = colorSet[index - 1];
  } else if (typeof colorSet === 'object') {
    color = colorSet[index];
  } else if (typeof colorSet === 'string') {
    color = colorSet;
  } else {
    color = name;
  }
  return color;
};

export const colorIsDark = (color) => {
  // https://stackoverflow.com/a/42429333
  const [red, green, blue] = color.match(/[A-Za-z0-9]{2}/g).map(v => parseInt(v, 16));
  // http://www.had2know.com/technology/
  //  color-contrast-calculator-web-design.html
  const brightness = (
    (299 * red) + (587 * green) + (114 * blue)
  ) / 1000;
  return (brightness < 125);
};

export default { colorForName, colorIsDark };
