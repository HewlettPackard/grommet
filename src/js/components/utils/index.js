export * from './styles';
export * from './mixins';
export * from './DOM';
export { default as Drop } from './Drop';
