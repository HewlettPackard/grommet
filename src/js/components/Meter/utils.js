import { colorForName } from '../../utils';

export const strokeProps = (color, theme) => {
  const result = {};
  if (color) {
    if (typeof color === 'object') {
      result.stroke = colorForName(color.color, theme);
      if (color.opacity) {
        result.strokeOpacity = (color.opacity === true ?
          theme.global.opacity.medium : theme.global.opacity[color.opacity]);
      }
    } else {
      result.stroke = colorForName(color, theme);
    }
  }
  return result;
};

export default { strokeProps };
