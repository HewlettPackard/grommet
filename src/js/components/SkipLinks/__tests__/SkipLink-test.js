import React from 'react';
import { Simulate } from 'react-dom/test-utils';
import 'jest-styled-components';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { createPortal, expectPortal } from '../../../utils/portal';

import { SkipLinks, SkipLink, SkipLinkTarget } from '../';

Enzyme.configure({ adapter: new Adapter() });

describe('SkipLink', () => {
  beforeEach(createPortal);

  test('mounts', (done) => {
    mount(
      <div>
        <SkipLinks id='skip-links'>
          <SkipLink id='main' label='Main Content' />
          <SkipLink id='footer' label='Footer' />
        </SkipLinks>
        <div>
          <SkipLinkTarget id='main' />
          Main Content
          <input type='text' value='main content' />
        </div>
        <footer>
          <SkipLinkTarget id='footer' />
          <input type='text' value='footer' />
        </footer>
      </div>
    );

    expectPortal('skip-links').toMatchSnapshot();

    document.getElementById('skip-links').querySelector('a').focus();
    expectPortal('skip-links').toMatchSnapshot();

    Simulate.click(document.activeElement);
    document.getElementById('skip-links').querySelector('a').blur();

    setTimeout(() => {
      expect(document.activeElement).toMatchSnapshot();
      done();
    }, 50);
  });
});
