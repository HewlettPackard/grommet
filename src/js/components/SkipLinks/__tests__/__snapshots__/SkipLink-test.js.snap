// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkipLink mounts 1`] = `
<div
  class="StyledLayer-ceZzeh bcWTAx"
  id="skip-links"
  tabindex="-1"
>
  <div
    class="StyledLayer__StyledContainer-gBGbkw buXZQX"
  >
    <div
      class="StyledBox-bStriS kLwTeJ"
      direction="column"
    >
      <h2
        class="StyledHeading-NQveS kJELbK"
      >
        Skip To
        :
      </h2>
      <div
        class="StyledBox-bStriS bqSFqK"
        direction="row"
      >
        <div
          class="StyledBox-bStriS cvUrkJ"
          direction="column"
        >
          <a
            class="StyledAnchor-hcnTIZ kwSIan"
            label="Main Content"
            href="#main"
          >
            Main Content
          </a>
        </div>
        <div
          class="StyledBox-bStriS cvUrkJ"
          direction="column"
        >
          <a
            class="StyledAnchor-hcnTIZ kwSIan"
            label="Footer"
            href="#footer"
          >
            Footer
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SkipLink mounts 2`] = `
"

.bcWTAx {
  font-family: 'Work Sans',Arial,sans-serif;
  font-size: 1em;
  line-height: 1.5;
  color: #333333;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  position: relative;
  z-index: 10;
  width: 100vw;
  height: 100vh;
  overflow: auto;
  background-color: rgba(0,0,0,0.5);
  left: -100%;
  right: 100%;
  z-index: -1;
  position: fixed;
}

@media only screen and (max-width:699px) {
  .buXZQX {
    min-height: 100%;
    min-width: 100%;
  }
}

@media only screen and (min-width:700px) {
  .buXZQX {
    position: absolute;
    max-height: 100%;
    max-width: 100%;
    overflow: auto;
    border-radius: 4px;
  }
}"
`;

exports[`SkipLink mounts 3`] = `
<div
  class="StyledLayer-ceZzeh diSzRA"
  id="skip-links"
  tabindex="-1"
>
  <div
    class="StyledLayer__StyledContainer-gBGbkw dLSDjx"
  >
    <div
      class="StyledBox-bStriS kLwTeJ"
      direction="column"
    >
      <h2
        class="StyledHeading-NQveS kJELbK"
      >
        Skip To
        :
      </h2>
      <div
        class="StyledBox-bStriS bqSFqK"
        direction="row"
      >
        <div
          class="StyledBox-bStriS cvUrkJ"
          direction="column"
        >
          <a
            class="StyledAnchor-hcnTIZ gUtwfp"
            label="Main Content"
            href="#main"
          >
            Main Content
          </a>
        </div>
        <div
          class="StyledBox-bStriS cvUrkJ"
          direction="column"
        >
          <a
            class="StyledAnchor-hcnTIZ kwSIan"
            label="Footer"
            href="#footer"
          >
            Footer
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SkipLink mounts 4`] = `
"

.diSzRA {
  font-family: 'Work Sans',Arial,sans-serif;
  font-size: 1em;
  line-height: 1.5;
  color: #333333;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  position: relative;
  z-index: 10;
  width: 100vw;
  height: 100vh;
  overflow: auto;
  background-color: rgba(0,0,0,0.5);
}

@media only screen and (min-width:700px) {
  .diSzRA {
    position: fixed;
    top: 0px;
    left: 0px;
    right: 0px;
    bottom: 0px;
  }
}

@media only screen and (max-width:699px) {
  .buXZQX {
    min-height: 100%;
    min-width: 100%;
  }
}

@media only screen and (min-width:700px) {
  .buXZQX {
    position: absolute;
    max-height: 100%;
    max-width: 100%;
    overflow: auto;
    border-radius: 4px;
  }
}

@media only screen and (max-width:699px) {
  .dLSDjx {
    min-height: 100%;
    min-width: 100%;
  }
}

@media only screen and (min-width:700px) {
  .dLSDjx {
    position: absolute;
    max-height: 100%;
    max-width: 100%;
    overflow: auto;
    border-radius: 4px;
    top: 0;
    left: 50%;
    -webkit-transform: translate(-50%,0);
    -ms-transform: translate(-50%,0);
    transform: translate(-50%,0);
    -webkit-animation: bAkQbM 0.2s ease-in-out forwards;
    animation: bAkQbM 0.2s ease-in-out forwards;
  }
}"
`;

exports[`SkipLink mounts 5`] = `
<body
  style="overflow: scroll;"
>
  <div>
    <div
      hidden=""
    >
      <div
        class="StyledLayer-ceZzeh bcWTAx"
        id="skip-links"
        tabindex="-1"
      >
        <div
          class="StyledLayer__StyledContainer-gBGbkw buXZQX"
        >
          <div
            class="StyledBox-bStriS kLwTeJ"
            direction="column"
          >
            <h2
              class="StyledHeading-NQveS kJELbK"
            >
              Skip To
              :
            </h2>
            <div
              class="StyledBox-bStriS bqSFqK"
              direction="row"
            >
              <div
                class="StyledBox-bStriS cvUrkJ"
                direction="column"
              >
                <a
                  class="StyledAnchor-hcnTIZ kwSIan"
                  label="Main Content"
                  href="#main"
                >
                  Main Content
                </a>
              </div>
              <div
                class="StyledBox-bStriS cvUrkJ"
                direction="column"
              >
                <a
                  class="StyledAnchor-hcnTIZ kwSIan"
                  label="Footer"
                  href="#footer"
                >
                  Footer
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-hidden="false"
  />
</body>
`;
